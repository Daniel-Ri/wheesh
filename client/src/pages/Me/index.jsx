import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { createStructuredSelector } from 'reselect';
import { selectUser } from '@containers/Client/selectors';
import RecentActorsOutlinedIcon from '@mui/icons-material/RecentActorsOutlined';
import KeyIcon from '@mui/icons-material/Key';
import MailIcon from '@mui/icons-material/Mail';

import { Avatar, Button } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { logout } from '@containers/Client/actions';
import toast from 'react-hot-toast';
import classes from './style.module.scss';

const Me = ({ user }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const clickHeader = () => {
    if (user) {
      navigate('/profile');
    } else {
      navigate('/login');
    }
  };

  const clickPassengerPage = () => {
    if (!user) {
      toast.error('Need to login');
    } else {
      navigate('/myPassengers');
    }
  };

  const clickChangePasswordPage = () => {
    if (!user) {
      toast.error('Need to login');
    } else {
      navigate('/changePassword');
    }
  };

  const clickChangeEmailPage = () => {
    if (!user) {
      toast.error('Need to login');
    } else {
      navigate('/changeEmail');
    }
  };

  return (
    <main className={classes.main}>
      <div className={classes.container}>
        <header className={classes.header} onClick={clickHeader}>
          <Avatar className={classes.icon} />
          <h1>{user && user.username ? user.username : 'Login'}</h1>
        </header>

        <section>
          <h2>Informasi Akun</h2>
          <hr />
          <div className={classes.sectionDesc}>
            <div className={classes.links}>
              <div className={classes.link} onClick={clickPassengerPage}>
                <RecentActorsOutlinedIcon className={classes.icon} />
                <div className={classes.namePage}>Daftar Penumpang</div>
              </div>
              <div className={classes.link} onClick={clickChangePasswordPage}>
                <KeyIcon className={classes.icon} />
                <div className={classes.namePage}>Kata Sandi</div>
              </div>
              <div className={classes.link} onClick={clickChangeEmailPage}>
                <MailIcon className={classes.icon} />
                <div className={classes.namePage}>E-mail</div>
              </div>
            </div>
          </div>
        </section>

        {user && (
          <section>
            <Button onClick={() => dispatch(logout())}>Logout</Button>
          </section>
        )}
      </div>
    </main>
  );
};

Me.propTypes = {
  user: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  user: selectUser,
});

export default connect(mapStateToProps)(Me);
