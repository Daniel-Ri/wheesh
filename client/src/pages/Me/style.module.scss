@import '@styles/base';

// Reset
* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: Poppins;
}

.main {
  margin-left: 200px;
  position: relative;
  background: $color-alabaster;
  display: flex;
  justify-content: center;
  width: calc(100vw - 200px);
  height: 100%;
  min-height: 100vh;

  .container {
    width: 90%;
    max-width: 40rem;
    @include directionY;

    header {
      margin-top: 5vh;
      @include directionX;
      gap: 1rem;
      align-items: center;
      cursor: pointer;

      .icon {
        color: $color-picton-blue-dark-shade-2;
        background-color: $color-light-blue;
      }

      h1 {
        font-size: 160%;
      }
    }

    section {
      margin-top: 1rem;
      @include directionY;
      box-shadow: 0 0 5px 0 $color-cool-gray;
      border-radius: 10px;

      h2 {
        padding: 0.75rem;
        font-size: 120%;
      }

      hr {
        border: 0;
        border-top: 1px solid $color-cool-gray;
      }

      .sectionDesc {
        width: 100%;
        padding: 1rem;

        .links {
          @include directionX;
          flex-wrap: wrap;
          gap: 1rem 0;

          .link {
            flex: 50%;
            @include directionX;
            gap: 0.5rem;
            align-items: center;
            cursor: pointer;

            .icon {
              color: $color-picton-blue-dark-shade-2;
              background-color: $color-light-blue;
            }

            .namePage {
              font-size: 100%;
            }
          }
        }
      }
    }
  }

  @media screen and (max-width: $screen-lg-limit) {
    margin-left: 192px;
    width: calc(100vw - 192px);
  }

  @media screen and (max-width: $screen-md-limit) {
    margin-left: 10rem;
    width: calc(100vw - 10rem);
  }

  @media screen and (max-width: $screen-sm-limit) {
    margin-left: 0;
    width: 100vw;
    margin-bottom: 76px;
    height: 100%;
    min-height: calc(100vh - 76px);

    .container {
      section {
        .sectionDesc {
          .links {
            .link {
              flex: 33%;
              @include directionY;
              gap: 0.1rem;

              .namePage {
                text-align: center;
                font-size: 80%;
              }
            }
          }
        }
      }
    }
  }
}