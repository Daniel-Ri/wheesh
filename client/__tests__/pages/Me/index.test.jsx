import { render, fireEvent } from '@utils/testHelper';
import Me from '@pages/Me';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

describe('Me Page', () => {
  test('Renders Me page with user not login', () => {
    const wrapper = render(<Me />);
    const { getByTestId, queryByTestId } = wrapper;
    expect(getByTestId('Me')).toBeInTheDocument();
    expect(getByTestId('MyPassengersLink')).toBeInTheDocument();
    expect(getByTestId('ChangePasswordLink')).toBeInTheDocument();
    expect(getByTestId('ChangeEmailLink')).toBeInTheDocument();
    expect(queryByTestId('WebManagement')).toBeNull();
  });

  test('Redirect to Login when not login', () => {
    const wrapper = render(<Me />);
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('Header'));
    expect(mockNavigate).toHaveBeenCalledWith('/login');
  });

  test('Cannot Redirect to MyPassengers when not login', () => {
    const wrapper = render(<Me />);
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('MyPassengersLink'));
    expect(mockNavigate).not.toHaveBeenCalledWith('/myPassengers');
  });

  test('Cannot Redirect to ChangePassword when not login', () => {
    const wrapper = render(<Me />);
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('ChangePasswordLink'));
    expect(mockNavigate).not.toHaveBeenCalledWith('/changePassword');
  });

  test('Cannot Redirect to ChangeEmail when not login', () => {
    const wrapper = render(<Me />);
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('ChangeEmailLink'));
    expect(mockNavigate).not.toHaveBeenCalledWith('/changeEmail');
  });

  test('Should match snapshot', () => {
    const wrapper = render(<Me />);
    expect(wrapper).toMatchSnapshot();
  });
});
