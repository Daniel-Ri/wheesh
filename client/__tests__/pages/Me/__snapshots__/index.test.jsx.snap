// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Me Page Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="Me"
      >
        <div
          class="container"
        >
          <header
            class="header"
            data-testid="Header"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault icon css-2s90m6-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                data-testid="PersonIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
            <h1>
              Login
            </h1>
          </header>
          <section>
            <h2>
              Informasi Akun
            </h2>
            <hr />
            <div
              class="sectionDesc"
            >
              <div
                class="links"
              >
                <div
                  class="link"
                  data-testid="MyPassengersLink"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                    data-testid="RecentActorsOutlinedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 5h2v14h-2zm-4 0h2v14h-2zm-3 0H2c-.55 0-1 .45-1 1v12c0 .55.45 1 1 1h12c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-1 12H3V7h10v10z"
                    />
                    <circle
                      cx="8"
                      cy="9.94"
                      r="1.95"
                    />
                    <path
                      d="M11.89 15.35c0-1.3-2.59-1.95-3.89-1.95s-3.89.65-3.89 1.95V16h7.78v-.65z"
                    />
                  </svg>
                  <div
                    class="namePage"
                  >
                    Daftar Penumpang
                  </div>
                </div>
                <div
                  class="link"
                  data-testid="ChangePasswordLink"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                    />
                  </svg>
                  <div
                    class="namePage"
                  >
                    Password
                  </div>
                </div>
                <div
                  class="link"
                  data-testid="ChangeEmailLink"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                    data-testid="MailIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
                    />
                  </svg>
                  <div
                    class="namePage"
                  >
                    E-mail
                  </div>
                </div>
              </div>
            </div>
          </section>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="Me"
    >
      <div
        class="container"
      >
        <header
          class="header"
          data-testid="Header"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault icon css-2s90m6-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
          <h1>
            Login
          </h1>
        </header>
        <section>
          <h2>
            Informasi Akun
          </h2>
          <hr />
          <div
            class="sectionDesc"
          >
            <div
              class="links"
            >
              <div
                class="link"
                data-testid="MyPassengersLink"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                  data-testid="RecentActorsOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 5h2v14h-2zm-4 0h2v14h-2zm-3 0H2c-.55 0-1 .45-1 1v12c0 .55.45 1 1 1h12c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-1 12H3V7h10v10z"
                  />
                  <circle
                    cx="8"
                    cy="9.94"
                    r="1.95"
                  />
                  <path
                    d="M11.89 15.35c0-1.3-2.59-1.95-3.89-1.95s-3.89.65-3.89 1.95V16h7.78v-.65z"
                  />
                </svg>
                <div
                  class="namePage"
                >
                  Daftar Penumpang
                </div>
              </div>
              <div
                class="link"
                data-testid="ChangePasswordLink"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                  />
                </svg>
                <div
                  class="namePage"
                >
                  Password
                </div>
              </div>
              <div
                class="link"
                data-testid="ChangeEmailLink"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                  data-testid="MailIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
                  />
                </svg>
                <div
                  class="namePage"
                >
                  E-mail
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
